New IA: Islamic Prayer Times

ozdemirburak opened the pull request:
**What does your Instant Answer do?**
Shows Islamic prayer times according to given or user's current location and time

**What problem does your Instant Answer solve (Why is it better than organic links)?**
Lets users to find the today's prayer times without providing location and time instead of searching within the list of whole locations, dates and times.

**What is the data source for your Instant Answer? (Provide a link if possible)**
~~http://www.aladhan.com/rest-api~~
http://www.muslimsalat.com/api/

**Why did you choose this data source?**
~~It's the most reliable one within the public ones.~~
It lets you search with location name, instead of providing both latitude and longitude.

**Are there any other alternative (better) data sources?**
~~Actually there is, where you can get the results without providing latitude and longitude, instead providing the city's name, which may expand this instant answer. However, within that data source,  after 25K requests, which is insufficient I guess, fee is needed for further requests. But maybe if we contact with them, they may grant unlimited requests.~~ 

**What are some example queries that trigger this Instant Answer?**
~~There are only three queries that trigger this Instant Answer which are 'namaz times', 'salat times' and 'islamic prayer times'. Maybe they can be extended, I may not be aware about all of the Islamic search queries about prayer times.~~
`islamic prayer times`, `izmir turkey namaz times`, `salat times in dc`, `namaz times for san marino`

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
Muslim people

**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**
https://duck.co/ideas/idea/971/the-daily-time-for-the-five-daily-islamic-prayers
https://duck.co/ideas/idea/6241/islamic-prayer-time

**Which existing Instant Answers will this one supersede/overlap with?**
None

**Are you having any problems? Do you need our help with anything?**
Currently nope. But I wanna ask you a question, if there exists a DDG GeoLocation service which returns latitude and longitude for given location. I've searched through the documentations and forums, but I might have missed it. 

**What are the terms of use for the API? Will DuckDuckGo need specific authorization (e.g. an API key)? Are there any costs associated with API usage?**
~~Open for public use. There are not any costs.~~
It needs an API key. It's free for 25K daily requests, for more, extra fee is needed. Details can be seen in [here](http://www.muslimsalat.com/panel/signup.php).

**Where did you hear about DuckDuckHack? (For first time contributors)**
Twitter, DDG official primary account.

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**

~~**With the old API**:~~

![islamic-prayer-times-ss1](https://cloud.githubusercontent.com/assets/5355510/9225304/554d2086-4111-11e5-9e2a-5d020d916b1e.png)

~~**New API (info):**~~

![islamic-prayer-times-ss4](https://cloud.githubusercontent.com/assets/5355510/9266355/7dc06b12-4248-11e5-80ea-b2de48cf809d.png)

**New API (list):**
![islamic-prayer-times-ss10](https://cloud.githubusercontent.com/assets/5355510/9366810/36d96ca2-46c2-11e5-975f-3c244c77e2f2.png)

## Checklist
Please place an 'X' where appropriate.

```
[x] Added metadata and attribution information
[x] Wrote test file and added to t/ directory
[x] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
[x] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
[x] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Ubuntu 14.04
        [x] Google Chrome
        [x] Firefox
        [] Opera

    - Windows 8
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 10

    - Windows 7
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 8
        [] IE 9
        [] IE 10

    - Windows XP
        [] IE 8

    - Mac OSX
        [] Google Chrome
        [] Firefox
        [] Opera
        [] Safari

    - iOS (iPhone)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - iOS (iPad)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - Android
        [] Firefox
        [] Native Browser
        [] Google Chrome
        [] Opera
```

https://duck.co/ia/view/islamic_prayer_times

moollaza commented:
@ozdemirburak thanks a lot for contributing, we really appreciate it! We'll be sure to review this shortly and give you some feedback!

>Actually there is, where you can get the results without providing latitude and longitude, instead providing the city's name, which may expand this instant answer. However, within that data source, after 25K requests, which is insufficient I guess, fee is needed for further requests. But maybe if we contact with them, they may grant unlimited requests. Can be seen in http://www.muslimsalat.com/api/

I think this would be a better choice ultimately. Do you know if they have support for specific locations, or the whole world? It seems like the latter, but I can't tell for sure.

>Currently nope. But I wanna ask you a question, if there exists a DDG GeoLocation service which returns latitude and longitude for given location. I've searched through the documentations and forums, but I might have missed it.

Unfortunately we don't have anything like this yet, but it's on our radar and I personally can't wait to have this functionality! It would certainly be useful for many IA's.

moollaza commented:
@abeyang any thoughts on design here? Seeing as the prayer times are static, I was wondering if we should opt to show tiles for the prayers of the day, and autoscroll to the tile for the next coming prayer?

ozdemirburak commented:
@moollaza I agree with you on the API choice and I also guess that it's for the whole world as they are probably using some service to obtain latitude and longitude from city, country or state. Should I better contact with the creators of the API ?

moollaza commented:
>Should I better contact with the creators of the API ?

@ozdemirburak if you wouldn't mind, that would be awesome. Many API providers have let us use their services for free because we clearly attribute them on the page. 

Is the 25K limit daily? Monthly? If they don't let us use it for free, we can go with the free plan for now and if we get close to the limit, we can get in touch with them again and see if we can partner.

ozdemirburak commented:
@moollaza It's [daily](http://www.muslimsalat.com/panel/signup.php). I've sent an email about the questions, if I got an answer, I'll post it here.

ozdemirburak commented:
@moollaza I've updated the code and proposal since I've changed the API as the previous one was limiting the scope of this IA. However, I haven't received any answers upon my email that I've sent to the API creators about the daily limits.

Note: I've created an test account for the API, tested with many locations and I can comfortably tell you that it's worldwide.

moollaza commented:
>It's daily. I've sent an email about the questions, if I got an answer, I'll post it here.

Ah daily is perfect. I don't think we risk hitting that right now, especially if we cache the results for a bit. Hopefully we'll hear back from them soon!

> I've created an test account for the API, tested with many locations and I can comfortably tell you that it's worldwide.

@ozdemirburak excellent, thanks a lot!

ozdemirburak commented:
@moollaza I've made some changes as you can see from the latest commit. However I have a another question. Let's say user is located in New York and searching for London prayer times, should we show the `getClosest()` result or not. Because it's currently showing the difference between user's local time, which is New York time, and the closest prayer time from London, so it's a little bit weird, but I'm not sure about how we should handle it.

moollaza commented:
@ozdemirburak are you asking if we should show the prayer time relative to the user in NY, vs the localtime in London? If so, it should be London local time.

That's how we tend to handle things. The sunrise IA is an example: https://duckduckgo.com/?q=sunrise+australia&t=ffab&ia=answer

ozdemirburak commented:
@moollaza Actually it's showing in London local time in the example that I've provided. But my question is something different.

 `getClosest()` returns the time difference between user's current local time and the closest prayer time from given location.

To give more accurate example, I'm changing the locations. Let's say, when the user from Izmir, Turkey searches for `salat times izmir turkey` right now, it will return `Fajr, in 5 hours` where Fajr is at 4:43 am tomorrow, in Izmir local time. However when the same user searches for `sydney salat times`, it will return `Fajr, in 5 hours.` where Fajr is at 5:12 am, in Sydned local time.

ozdemirburak commented:
@moollaza Actually it's showing in London local time in the example that I've provided. But my question is something different.

 `getClosest()` returns the time difference between user's current local time and the closest prayer time from given location.

To give more accurate example, I'm changing the locations. 

Let's say, currently, user from Izmir, Turkey searches for `salat times izmir turkey`, it will return `Fajr, in 5 hours` where Fajr is at 4:43 am tomorrow, in Izmir local time. However when the same user searches for `sydney salat times`, it will return `Fajr, in 5 hours.` where Fajr is at 5:12 am, in Sydney local time. 

However, it's already 7:10 AM in Sydney local time and 00:10 in Izmir local time and Fajr event is already past in Sydney and it's not `in 5 hours` for the user who is in Izmir. 

So, I'm asking that should we add the timezone offset for calculating the closest pray time, or completely remove the closest thing when user searches for another location?

moollaza commented:
Ah, I think I understand. This is mostly a design problem, so I'd like @abeyang thoughts, but I think we should focus on providing the exact local time of the next prayer. And if we have the current time in that location,  indicate the relative time until Salat in terms of local time. 

Trying to convert and account for the time difference is messy and confusing IMO. 

Does that make sense?

ozdemirburak commented:
@moollaza Now, showing all the differences according to local time. I think screenshots can be more illuminating.

![izmir](https://cloud.githubusercontent.com/assets/5355510/9266231/e4978aa2-4246-11e5-9bc4-0e7d8bea444b.png)
![nyc](https://cloud.githubusercontent.com/assets/5355510/9266232/e49aed50-4246-11e5-8c14-96381575fcb8.png)
![dakar](https://cloud.githubusercontent.com/assets/5355510/9266230/e496f8bc-4246-11e5-964c-ade41683ef50.png)

jagtalon commented on share/spice/islamic_prayer_times/islamic_prayer_times.js:
```diff
@@ -0,0 +1,93 @@
+
+        DDG.require('moment.js', function() {
+
+            function capitalize(string) {
```
You can use the built in function `DDG.capitalize`

jagtalon commented:
Just created an IA page for this: https://duck.co/ia/view/islamic_prayer_times

ozdemirburak commented:
@jagtalon Thanks for the feedback, updated.

moollaza commented on share/spice/islamic_prayer_times/islamic_prayer_times.js:
```diff
@@ -0,0 +1,89 @@
+                    group: 'info',
+                    options: {
+                        content: Spice.islamic_prayer_times.islamic_prayer_times,
+                        moreAt: false
```
Can we possibly construct and send them to a link like http://muslimsalat.com/london/daily for more info?

moollaza commented on share/spice/islamic_prayer_times/islamic_prayer_times.js:
```diff
@@ -0,0 +1,89 @@
+                templates: {
+                    group: 'info',
+                    options: {
+                        content: Spice.islamic_prayer_times.islamic_prayer_times,
```
nit: the template file should be named `content.handlebars` so its referenced as `Spice.islamic_prayer_times.content`

moollaza commented:
@ozdemirburak after speaking with the design team we agreed we have so great ideas for a possible V2 of this IA, but we'd like to simplify it for now and get it live.

Let's switch this to the `List` template and simply show a list of the prayer times for the given location. I think we can stick to a simple header that says `<location> on <date>`. Something like this should work:

> **Next prayer: Fajr at 4:40AM, Aug 14 (Tomorrow)**
> New York, USA 
>
> -----
> Fajr: 4:56 AM

> -----
> Sunrise: 6:23 AM

> -----
> Dhuhr: 1:20 PM

> -----
> Asr: 5:11 PM

> -----
> Maghrib: 8:16 PM

> -----
> Isha: 9:44 PM


/cc @abeyang how does that sound?

ozdemirburak commented:
@moollaza Updated. This is how it looks like now:

![islamic-prayer-times-ss5](https://cloud.githubusercontent.com/assets/5355510/9349778/2f0866b4-4653-11e5-966b-a7bc88f97abc.png)

And for the tomorrow case, its' header will look like this

![islamic-prayer-times-ss6](https://cloud.githubusercontent.com/assets/5355510/9349789/4c2c27e4-4653-11e5-983d-29e6815f547d.png)

abeyang commented:
@ozdemirburak Looking good! Few comments:

1. Can we put the info (prayer and time) into two columns? Similar to the [whois IA](https://duckduckgo.com/?q=whois+facebook.com&ia=whois).
2. Let's change the title/subtitle to this:

> **Fajr at 4:40AM, Aug 14 (Tomorrow)**
> Next prayer time at Phoenixville, PA

Note: For any US locations, let's lose the country ("USA") altogether. Thanks!

ozdemirburak commented:
@abeyang Updated.

US:
![islamic-prayer-times-ss7](https://cloud.githubusercontent.com/assets/5355510/9366651/448a7d38-46c1-11e5-856d-741b61450adb.png)

Non US:
![islamic-prayer-times-ss8](https://cloud.githubusercontent.com/assets/5355510/9366708/81579aac-46c1-11e5-9a5e-1f64090fcd22.png)

abeyang commented:
:+1:

moollaza commented on share/spice/islamic_prayer_times/islamic_prayer_times.js:
```diff
@@ -0,0 +1,97 @@
+
+        DDG.require('moment.js', function() {
+
+            function setDst(time, dst) {
```
Let's move these functions out of the callback. They can be moved above `env.ddg_spice_islamic_prayer_times`.

We call them from inside the closure, where `moment` is defined, so its OK :+1:

moollaza commented on share/spice/islamic_prayer_times/islamic_prayer_times.js:
```diff
@@ -0,0 +1,97 @@
+                data: api_result,
+                meta: {
+                    sourceName: "Muslim Salat",
+                    sourceUrl: api_result.link
```
Let's prepend "https://" here

ozdemirburak commented:
@moollaza Updated.

moollaza commented on share/spice/islamic_prayer_times/islamic_prayer_times.js:
```diff
@@ -0,0 +1,98 @@
+        return clone;
+    }
+
+    function getClosest(timings, offset, local) {
```
I think this code warrant some comments to explain what these are for and why we need them.

moollaza commented on lib/DDG/Spice/IslamicPrayerTimes.pm:
```diff
@@ -0,0 +1,52 @@
+            web     => ["http://burakozdemir.co.uk", "Burak Özdemir"];
+
+spice to => 'http://muslimsalat.com/$1/weekly.json?key={{ENV{DDG_SPICE_MUSLIMSALAT_APIKEY}}}';
+spice wrap_jsonp_callback => 1;
```
Looks like the API supports a `callback` param? Let's add `&callback={{callback}}` to `spice to`

moollaza commented:
@ozdemirburak this is just about done! There's just a couple small things to add and I'll be happy to merge this. Great job implementing this! Thanks a lot for contributing to DuckDuckHack!

ozdemirburak commented:
@moollaza Updated. Thanks, I should also note that I'm glad that I'm contributing to DuckDuckHack. I hope that I'll contribute more in the future.

moollaza commented on share/spice/islamic_prayer_times/islamic_prayer_times.js:
```diff
@@ -0,0 +1,100 @@
+        return index === 0 ? closest.format('h:mm A') : closest.format('h:mm A, MMM D ') + '(Tomorrow)';
+    }
+
+    function setAddress(title, country_code, country, state) {
```
Looks like there's a bug in here? I think we should always return the `city` + `country` and in the case of US, `city` + `state`?

Right now it looks like we're skipping the City in all cases, and returning `state` + `country` for US:

![namaz_times_for_new_york_at_duckduckgo](https://cloud.githubusercontent.com/assets/873785/9392208/53c11090-4749-11e5-9fcd-1fa4eaf98bb1.png)
![namaz_times_for_cairo_at_duckduckgo](https://cloud.githubusercontent.com/assets/873785/9392216/6143ba7e-4749-11e5-9340-b5519a60610e.png)

moollaza replied:
Hmm looks like the API has some issues with Unicode...
![namaz_times_for_tokyo_at_duckduckgo](https://cloud.githubusercontent.com/assets/873785/9392318/08c04de4-474a-11e5-828b-6f38e526f108.png)

That's a tough case because the city, state, country are all blank and there's no way for use to know if the unicode is correct.

moollaza commented:
@ozdemirburak one more small fix and then this should be good to merge :+1:

ozdemirburak commented:
@moollaza Updated. However, some places don't have the city names, or state names, or addresses. So I've checked the all, if you try Cairo again, it will return the same as nor city name neither address is not returned by the API.

I should also note that, some places give error because of the BIGINT scopes, as the pray times are returned as -9223372036854775808, which is the minimum within the BIGINT scope, from the API, try Istanbul for instance ( actually, Istanbul itself is a state as it's really big place, divided to many cities, some must make the search as "Şişli İstanbul" instead, but people may not think about it ). So I've added an error check for that and pushed into list if they are valid.

moollaza commented:
@ozdemirburak thanks a lot! This LGTM :+1: 

I noticed on more oddity which is a little difficult to fix. If you specify a state as the input the only part of the adress we get back is the state 2 letter code so "namaz times in California" shows "... at CA".

We may need to use the existing capital cities hash (see the Time Spice) and if the query matches a country name or state name, we may have to alter the API call to specify a specific city so it's clear where the result is coming from.

ozdemirburak commented:
@moollaza I'll be working on that issue within the next week.

Also, after the DDG location service is completed, this one can even be turned into a goodie, as the times can be calculated easily ( Can be seen [here](http://praytimes.org/calculation/) ).
